inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,368 @@
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/intel-inb-manageability/intel-inb-manageability/inbm/dispatcher-agent/')
+import pytest
+from unittest.mock import patch, MagicMock
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationAddSourceParameters
+from unittest.mock import patch, mock_open
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationSourceList
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationRemoveSourceParameters, SourceError, LINUX_GPG_KEY_PATH, UBUNTU_APT_SOURCES_LIST_D
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, SourceError, ApplicationUpdateSourceParameters
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceError
+from unittest.mock import patch, mock_open, call
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerAdd:
+    @pytest.fixture
+    def mock_dispatcher_broker(self):
+        return MagicMock()
+
+    @pytest.fixture
+    def mock_application_source_manager(self, mock_dispatcher_broker):
+        return UbuntuApplicationSourceManager(mock_dispatcher_broker)
+
+    @pytest.mark.parametrize('gpg_key_name, gpg_key_uri, source_list_file_name, sources', [
+        ('key1', 'http://example.com/key1', 'file1', ['deb http://example.com/ubuntu/ bionic main restricted']),
+        (None, None, 'file2', ['deb http://example.com/ubuntu/ bionic main restricted']),
+    ])
+    @patch('dispatcher.source.ubuntu_source_manager.verify_source')
+    @patch('dispatcher.source.ubuntu_source_manager.add_gpg_key')
+    @patch('dispatcher.source.ubuntu_source_manager.create_file_with_contents')
+    def test_add(self, mock_create_file_with_contents, mock_add_gpg_key, mock_verify_source, mock_application_source_manager, gpg_key_name, gpg_key_uri, source_list_file_name, sources):
+        # Arrange
+        parameters = ApplicationAddSourceParameters(gpg_key_name=gpg_key_name, gpg_key_uri=gpg_key_uri, source_list_file_name=source_list_file_name, sources=sources)
+
+        # Act
+        mock_application_source_manager.add(parameters)
+
+        # Assert
+        if gpg_key_name and gpg_key_uri:
+            mock_verify_source.assert_called_once_with(source=gpg_key_uri.rsplit('/', 1)[0], dispatcher_broker=mock_application_source_manager._dispatcher_broker)
+            mock_add_gpg_key.assert_called_once_with(gpg_key_uri, gpg_key_name)
+        mock_create_file_with_contents.assert_called_once_with(f'/etc/apt/sources.list.d/{source_list_file_name}', sources)
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerList:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)

<h4>For feedback purposes, please include both emoticon-based and qualitative feedback in your reply to this comment.</h4>

To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)